### PR TITLE
Add support for collection literals in the b6 shell and Python.

### LINKED ISSUE #66: Support lists as parameters to b6 functions
There are many situations in which it'd be useful to pass a hand-curated list to b6 functions - for example - to form the convex hull of a manually curated list of neighbouring boundary features, or to manually list a small number of important service providers (eg major hospitals). We neither support this in the shell or Python.

Reported by @gacafe

### PR DESCRIPTION
**Pull request by andreweland**
Add supprt for collection literals in the b6 shell and Python. In the shell, you can say:
{"primary", "secondary"}
for a collection with implicit integer keys from the value's index in the literal, or:
{highway="motorway": 2, highway="primary": 6}
for a collection with explicit keys and values (here tags and integers)

Similarly in Python, lists and dicts are converted into collections. Keys and values are evaluated server side, and hence can be arbitrary b6 expressions, not just literals.

Internally, this is implemented as a new b6 function, collection, which takes a variable number of pairs representing
individual keys and values, and turns them into a collection. This required added support in the b6 VM for functions
that have a variable number of arguments (variadic, in go speak), which is done using go's native support for variadic
functions. As variadic arguments in go need to be at the end of the argument list, this necessitates moving Context
from the end to the start in all b6 functions.

Fixes #66

### CODE REVIEW

File: src/diagonal.works/b6/api/functions/math.go [hunk 1]
```diff
@@ -7,35 +7,41 @@ import (
...
-	case api.FloatNumber:
-		switch a := a.(type) {
-		case api.IntNumber:
-			return api.FloatNumber(float64(a) / float64(b)), nil
-		case api.FloatNumber:
-			return api.FloatNumber(float64(a) / float64(b)), nil
 		}
+		return api.FloatNumber(float64(a) / float64(b.(api.FloatNumber))), nil
+	}
+	if b, ok := b.(api.IntNumber); ok {
```
Comment (xprogramari) [inline src/diagonal.works/b6/api/functions/math.go:17]: float number ?

Comment (xprogramari) [reply src/diagonal.works/b6/api/functions/math.go:17]: if i m right and i caught this i should get a peer bonus or a beer at some point

File: src/diagonal.works/b6/api/shell.go [hunk 2]
```diff
@@ -566,6 +568,79 @@ func reduceSymbolsSymbols(ss []*pb.NodeProto, s *pb.NodeProto) []*pb.NodeProto {
 	return reduced
 }
 
+func reduceCollectionItems(collection *pb.NodeProto) *pb.NodeProto {
```
Comment (xprogramari) [inline src/diagonal.works/b6/api/shell.go:571]: i think reduce is descriptive enough / but up to u

Comment (xprogramari) [reply src/diagonal.works/b6/api/shell.go:571]: but maybe u cant reuse that for other methods / idk how same name / different param works in go

File: src/diagonal.works/b6/api/shell.go [hunk 3]
```diff
@@ -566,6 +568,79 @@ func reduceSymbolsSymbols(ss []*pb.NodeProto, s *pb.NodeProto) []*pb.NodeProto {
...
+						Symbol: "pair",
+					},
+				},
+				Args: []*pb.NodeProto{key, value},
+			},
+		},
+	}
+}
+
+func reduceCollectionValueWithImplictKey(value *pb.NodeProto) *pb.NodeProto {
```
Comment (xprogramari) [inline src/diagonal.works/b6/api/shell.go:629]: does this make sense ? not passing a key ?

File: src/diagonal.works/b6/api/functions/math.go (hunk 1 above)
Comment (xprogramari) [reply src/diagonal.works/b6/api/functions/math.go:17]: ah nvm i see it, u re converting on b...no beer for me

File: src/diagonal.works/b6/api/shell.go (hunk 2 above)
Comment (andreweland) [reply src/diagonal.works/b6/api/shell.go:571]: Yeah, you can't reuse names at all, and "reduce" is the common term in a parser for for turning the right hand side of a production rule into the left, so it's used all over the place - hence I need to suffix everything.

File: src/diagonal.works/b6/api/shell.go (hunk 3 above)
Comment (andreweland) [reply src/diagonal.works/b6/api/shell.go:629]: I think it's useful? We use collections where the keys are just the index of each element within a collection (ie 0,2,3,4,5) in quite a few places, so this is just an easy way for specifying them in the shell (ie {"trunk", "primary"} evaluates to the same as {0: "trunk", 1: "primary"}. I guess we'll see if it actually gets used though!

Comment (xprogramari) [reply src/diagonal.works/b6/api/shell.go:629]: ah sorry, i turned the order upside down somehow, and thought keys were "trunk" / "primary"; that deffo makes sense

**Pull request merged**